### PR TITLE
Add partial support for LeakyRelu via "folding"

### PR DESCRIPTION
**Pull request by tlepley-cadence**
*Description*:

1) Added partial support for LeakyRelu:
- LeakyRelu Node in the IR
- LeakyRelu lowering for CPU/interpreter backends

2) Added an optimization to fold Max(Mul(Constant, A), A) to LeakyRelu (We've got some model where LeakyRelu is expressed this way).

From this PR, implementing LeakyRelu in importers will be simple for the day someone encounters directly the LeakyRelu operator in an ONNX or Caffe2 model.

For information, here is the ONNX LeakyRelu description:
https://github.com/onnx/onnx/blob/master/docs/Operators.md#LeakyRelu

*Testing*:
- Test node creation and lowering
- Test the folding optim

*Documentation*: N/A

### CODE REVIEW

File: lib/Optimizer/GraphOptimizer.cpp [hunk 1]
```diff
@@ -1345,6 +1345,64 @@ static void optimizeBatchNorm(Function *F) {
...
+/// Return true if the \p node sub-tree corresponds to a scalar
+/// Constant node and return the float value in \p retFloat.
+static bool getFloatScalarConstant(Node *node, float *retFloat) {
+  // Iterate accross potential Tile Nodes (implied by broadcasting if any).
+  auto *n = node;
+  while (auto *TN = dyn_cast<TileNode>(n)) {
+    n = TN->getInput();
+  }
+
+  // After potential Tile nodes, it should a singleton Constant node with any
```
Comment (artemrakhov) [inline lib/Optimizer/GraphOptimizer.cpp]:
```suggestion
  // After potential Tile nodes, it should be a singleton Constant node with any
```

File: lib/Optimizer/GraphOptimizer.cpp [hunk 2]
```diff
@@ -1345,6 +1345,64 @@ static void optimizeBatchNorm(Function *F) {
...
+  // shape corresponding to one single element.
+  if (Constant *constNode = dyn_cast<Constant>(n)) {
```
Comment (artemrakhov) [inline lib/Optimizer/GraphOptimizer.cpp]: Please do early return:
```suggestion
  Constant *constNode = dyn_cast<Constant>(n);
  if (!constNode || constNode->getType()->getElementType() != ElemKind::FloatTy
      || constNode->getType()->size() != 1)
    return false;
  ...
```

**Comment by tlepley-cadence** [comment]
@artemrakhov thanks for the feedback. It indeed works (only) because the fold optim matches Constant while lowering generated Splat

I was planning to file an issue to discuss this general folding vs lowering problematic in a second step, because I didn't find yet a convenient general solution and it goes beyond this PR. 

I'm also not convinced by the solution you propose, that consists of always forcing the lowering of the IR before the actual lowering stage. 

I think we should do the opposite: fold as much as possible before lowering:
- it would make the IR more cannonical and would simplify many optimizations. An other example is for instance BatchNorm that is often unfolded in the input model with many possible forms. 
- it would make the compiler flow more consistent, going from high level/high abstraction to low level/low abstraction, stage after stage.
- it would simplify backends. I don't think it's good to have a high level IR driven by the CPU backend specificity (many backends will support high level nodes natively).

Note that today, we have Relu in the high IR, so it loosk consistent to have LeakyRelu too.

A solution that looks better to me would be to split the optimizations in 2 parts:
1/ IR 'legalizer' called *only* after model import to make the high level IR more canonical. In particular, ONNX models that are converted from various frameworks and from various converter tools can use many different nodes to express the same functionality. 
2/ General optims (sink, merge etc...)

Then, we would have as compiler flow:

1- import model
2- legalize the IR (fold high level nodes as much as possible)
3- generic optims (no folding)
4- lowering
5- generic optims (no folding)

**Comment by tlepley-cadence** [comment]
@artemrakhov I've updated the PR, taking into account your code feedbacks, and implementing the 'legalization' approach that I propose in #2615 . Since there is no more problem of re-folding after lowering, I've made the folding optim for LeakyRelu more generic (handling both Splat and Constant). No more hack then :)

**Comment by artemrakhov** [comment]
I don't think that "legalize" is the right name. I see that there is "legalization" phase in LLVM, but it replaces unsupported instructions with supported instructions. In this PR, legalization does the opposite: replaces Max(x, x * alpha) _(supported by every backend)_ with LeakyReluNode _(not supported by CPU and Interpreter backends)_.

> I don't think the solution you propose (forcing the lowering of the IR before the actual lowering stage) is the right way to go, I think we should do the opposite: fold as much as possible before lowering.

I want to outline that this solution is backend-dependent. CPU and Interpreter will never fold, only lower. Backend XYZ with native support of LeakyRelu will disable lowering for this node, and enable folding as it's own Backend-specific code.
Please see how `CPUBackend::transformPostLowering` folds Conv and Relu for CPU.

**Comment by mortzur** [comment]
Hi @tlepley-cadence ,
I've started working on #2501 , which requires PRelu implementation.
PRelu is a parameterized leaky Relu , meaning it takes in a variable for the slope (instead of a constant).
I don't think there is anything specific we should do about this right now. I just wanted to draw your attention to this and to be in sync.  
I'll try and see if I can leverage the leaky-Relu implementation for this.

**Comment by artemrakhov** [comment]
PRelu in theory supersedes LeakyRelu. We should not have both. But we should go with different lowering formula: `Select(CmpLTENode(x, zeroSplat), Mul(x, alpha), x)`

@tlepley-cadence : I think your backend can do its own specific transformation in `transformPostLowering()`. It will match both `x < 0 ? ... ` and `max(x * alpha, x)` patterns and fold them to your backend's specific `LeakyReluNode`.

**Comment by tlepley-cadence** [comment]
@artemrakhov I don't think we can hide the folding of leakyRelu in our backend. We want these folding transformation to happen on the float model, not on a quantized model. Also, LeakyRelu has nothing specific to our backend. This is a fully standard operator of ONNX and Caffe2 that IMO has its place in the IR, like Relu. 

Regarding ops that supersed others, we can also argue that LeakyRelu superseds Relu. At the end, everything could be expressed as PRelu, but expressing something simple with something complex  will complicate optimizations, lowering,  backends. 

What I try to achieve here is to fold in the IR standard ONNX & Caffe2 operators (LeakyRelu in this particular case) that are described lowered in the input model for some reasons (usually, it's a conversion side effect, it's relatively common with ONNX). After model import (and legalization), I think the IR should correspond to the original operators used by the model. For instance, in the case of the yolo ONNX model we use, the node should be a LeakyRelu, not a PRelu.

**Comment by tlepley-cadence** [comment]
I renamed 'legalize' with 'fold' to avoid overloading a word meaning something else in the LLVM world. I propose to handle the discussion of the management of the various Relu variants separately (PRelu, LeakyRelu, Relu6, Relu etc..). 
Summary:
- today, there is only Relu in the IR and would need more support for LeakyRelu
- having operators lowered in the input model is a general issue we face. It's not only about Relu. @jfix71 mentionned LSTM. We also got issues with BatchNorm. We need these transformation to happen as early as possible in the compiler flow and at float level. Handling them late in the flow once the model is quantized brings complexity and risks.
- The operators we want to fold are 'legitimate' operators (defined in ONNX/Caffe2) and not backend specific. The choice of lowering them or not at the end is neverthelesss a backend specific choice.

In #2615 @jfix71 proposes some backend control for the folding pass, similarly to the control it has on the lowering process. @artemrakhov would this suit you ?

**Comment by nadavrot** [comment]
I don't think that Glow should Fold things by default. How do you guys feel about making this a utility function that backends can optionally use?

**Comment by tlepley-cadence** [comment]
@nadavrot if it's an opt-in, I'm fine with this. We could have a solution like proposed by @jfix71: the 'folding' pass I propose that happens after model import is controlled by backends (like the lowering pass). The folding of operators would be disabled by default and backend could enable it operators per operators (`shouldFold` virtual function).

**Comment by tlepley-cadence** [comment]
@nadavrot well, actually thinking more about it, it would not work for us. We use the CPU backend to generate the quantization profile. If the CPU backend and our backend behave differently before lowering, it will not work because the IR can be different and then the quantization profile may not be correct.

**Comment by tlepley-cadence** [comment]
Personally, I still don't see much the problem of folding operators after import. A well formed input model should have the higher level nodes. Then, with a good input model, the IR should already contain the high level node. What we try to handle here with this 'folding' pass is 'malformed' models  that are often the result of the conversion from a framework to ONNX. I don't really understand the problem of 'highering' the IR for these malformed models at the beginning of the Glow flow. Any backend still has control at the lowering stage.

**Comment by jfix71** [comment]
>If the CPU backend and our backend behave differently before lowering, it will not work because the IR can be different and then the quantization profile may not be correct.

One way we could solve this is we could keep track of what is folded into what, similar to `lower()` with `LoweredInfoMap`, and use it in the `FunctionQuantizer`.

So assume some folding `A->B->C` into `D`. The XBackend wants to do this folding, while the CPUBackend doesn't. You profile on the CPUBackend with `A->B->C`. Then XBackend would fold it after import/before quantization, and in `LoweredInfoMap` we'd log `{D: C}`. Then quantization would look for a profile for `D`, not find it, then check `LoweredInfoMap`, find `C`, and it has the profile it's looking for.

**Comment by tlepley-cadence** [comment]
In this PR, there were 2 features:
- Support for a LeakyRelu operator
- Folding of LeakyRelu

@artemrakhov @nadavrot I updated this branch to only keep the LeakyRelu operator support. The other item (folding of operators) looks controversial and I think that it's better to continue to discuss it separately (in #2615 for instance) without to delay the main purpose of this PR that is to add some support for LeakyRelu in the IR.

**Comment by artemrakhov** [comment]
**Now main question is:** How is `LeakyRelu` going to work for your Backend, if you use CPU to generate quantization profiles? `LeakyRelu` will be lowered for CPU, and you'll see profiles for individual primitives (`Max`, `Mul`). These profiles need to be merged together carefully to get profile for `LeakyRelu`. Do you have functionality to do this?

I'm OK with having "fold" or "raise" pass at early stage, if the motivation is to handle malformed models. Note, that Interpreter, CPU, OpenCL and many backends will prefer to lower as much as possible, so they won't want to "fold" or "raise" anything.

**Comment by tlepley-cadence** [comment]
@artemrakhov The generation of QuantizationProfile node were indded recently moved to after the lowering. The information of the link between the non lowered node and the lowerer node is kept (in LoweredInfoMap?). Then, Glow generates a quantization profile tjat contain both the non lowered and lowered nodes. If the other backend doesn't lower, while the CPU lowers, it still works.

**Comment by nadavrot** [comment]
@tlepley-cadence @artemrakhov I understand the problem now. I am okay with having a folding pass early in the compilation pipeline. It makes sense and can unblock optimizations.

File: lib/Optimizer/GraphOptimizer.cpp [hunk 3]
```diff
@@ -2453,6 +2453,86 @@ void glow::convertPlaceholdersToConstants(Function *F,
...
+  // Iterate accross potential Tile Nodes (implied by broadcasting if any).
+  auto *n = node;
+  while (auto *TN = dyn_cast<TileNode>(n)) {
+    n = TN->getInput();
+  }
+
+  // After potential Tile nodes, it should be a singleton constant scalar node
+  // with any shape corresponding to one single element.
+  if (auto *constNode = dyn_cast<Constant>(n)) {
+    if ((constNode->getType()->getElementType() != ElemKind::FloatTy) ||
```
Comment (opti-mix) [inline lib/Optimizer/GraphOptimizer.cpp]: We support multiple FP types, right? Float and FP16. We also have a predicate `isFPType` that can be used to check if it is one of those types.

**Comment by tlepley-cadence** [comment]
Thanks @nadavrot ! I 've re-updated the branch with the folding pass for leaky-relu.

Later, we can discuss other operation folding for the case we have detected. I'm thinking in particular to BatchNormalization that we usually want to be merged with convolution (thanks to the existing `optimizeBatchNorm` optimization). But some malformed ONNX models have it lowered.

File: lib/Optimizer/GraphOptimizer.cpp (hunk 1 above)
Comment (tlepley-cadence) [reply lib/Optimizer/GraphOptimizer.cpp]: Done

File: lib/Optimizer/GraphOptimizer.cpp (hunk 2 above)
Comment (tlepley-cadence) [reply lib/Optimizer/GraphOptimizer.cpp]: Done

**Comment by nadavrot** [comment]
@tlepley-cadence This looks very similar to #2619. Is it the same operator? Can we share the logic?
Let's split the PR and land the pieces one by one. I think that the folding code makes sense.

**Comment by tlepley-cadence** [comment]
@nadavrot We need to land the operator before the folding code. PRelu is significantly more complicated than LeakyRelu because it has Alpha given as a (potentially broadcasted) non constant input tensor while LeakyRely has a simple constant scalar Alpha parameter. IMO, LeakyRelu is closer to Relu than to PRelu: Relu is LeakyRelu with alpha == 0 and I would better see at some point LeakyRelu fused with Relu in the IR rather than with PRelu.

**Comment by jackm321** [comment]
I think the folding approach makes sense. The `sinkTranposeBelowChannelShuffle` optimization is an example where we basically do this implicitly but it would be cleaner with folding.

**Comment by nadavrot** [comment]
@artemrakhov @mortzur @tlepley-cadence How do you suggest that we move forward here? I don't think that we should have 3 operators. Considering that RELU is significantly more common, maybe we should merge LRelu with PRelu?  After all, we should be able to pattern match and optimize inside of the different backends for the special case where the PRelu parameter is a splat-broadcast constant. No?

**Comment by artemrakhov** [comment]
I suggest that we should have RELU and PRelu, as I mentioned before. It's possible to express LeakyRelu as PRelu. There's no need for LeakyRelu separate node in any of our backends. Then if some other backend wants to support LeakyRelu on low level, it can easily pattern-match case when "alpha" for PRelu is given as Splat.

I support adding folding pass, but it needs to fold to PRelu with Splat.

**Comment by tlepley-cadence** [comment]
With quantization, it's always a bit tricky and simple things can get complicated practically. I'm going to rapidly do some experiments with PRelu now that it has landed, and see if it can or not be a convenient replacement of the LeakyRelu operator with our target. For now, I'll keep this PR open.

**Comment by jfix71** [comment]
@tlepley-cadence I also want note that right now we are already calling `optimize()` and then `lower()` prior to `optimizeFunction()` (e.g. in `Loader::compile()`) We of course also call `lower()` after `fold()` too here in `optimizeFunction()`, meaning that we'd do `lower()`, `fold()`, `lower()` with this PR. I would like to go back to lowering a single time (https://github.com/pytorch/glow/issues/2648).

Either we should redesign the quantization and profiling transformations so that they occur inside `optimizeFunction()` after `lower()` (as noted in https://github.com/pytorch/glow/issues/2648), or otherwise perhaps move `fold()` to occur at the end of importing the model (e.g. at the end of the `setup()`  call in the `ONNXModelLoader` constructor), so that we're sure no other graph transformations have occurred before folding. Do you have any thoughts here?

**Comment by sergeigofman** [comment]
Need to be carefil with folding Max(x, x * alpha) to LeakyRelu/PRelu alpha is negative. I have encountered such models. Suggest to enhance the condition to guard against that.

**Comment by tlepley-cadence** [comment]
@jfix71 I had not noticed that `lower()` was called twice now. I guess it's related to the changes you did to get profiling information on lowered nodes but not in all cases (partial-lowering -> profile instrumentation -> full lowering). I then moved the call to `fold()` to `Loader` since the idea is to have a **single** folding pass just after model import.

**Comment by tlepley-cadence** [comment]
@artemrakhov @nadavrot I've updated the PR to fold the 'malformed' LeakyRelu pattern we have detected in some ONNX models as `PRelu(Splat)`. The PR then doesn't add anymore a new `LeakyRelu` operator, but achieves the same goal for us.

File: tests/unittests/GraphOptzTest.cpp [hunk 4]
```diff
@@ -1999,6 +2001,64 @@ TEST_F(GraphOptz, fusePadIntoConvNeg2) {
                       1 /* convStride */, 16 /* convNumKernels */);
 }
 
+TEST_F(GraphFold, foldLeakyReluFromSplat) {
+  // Max(A, Mult(A, Splat)) -> PRelu(Splat)
```
[the remark this answers is not in the document]
Comment (tlepley-cadence) [reply tests/unittests/GraphOptzTest.cpp]: Done.

File: tests/unittests/GraphOptzTest.cpp [hunk 5]
```diff
@@ -1999,6 +2001,64 @@ TEST_F(GraphOptz, fusePadIntoConvNeg2) {
...
+  auto *newPReluNode = llvm::dyn_cast<PReluNode>(output->getInput());
+  ASSERT_TRUE(newPReluNode);
+  auto *newSplatNode = llvm::dyn_cast<SplatNode>(newPReluNode->getSlope());
+  ASSERT_TRUE(newSplatNode);
+  EXPECT_EQ(leakyAlpha, newSplatNode->getValue());
+  EXPECT_EQ(input, newPReluNode->getInput());
+}
+
+TEST_F(GraphFold, foldLeakyReluFromConst) {
+  // Max(A, Mult(A,broadcasted Const)) -> PRelu(Splat)
```
[the remark this answers is not in the document]
Comment (tlepley-cadence) [reply tests/unittests/GraphOptzTest.cpp]: Done.

File: lib/Optimizer/GraphOptimizer.cpp [hunk 6]
```diff
@@ -2446,6 +2446,89 @@ void glow::convertPlaceholdersToConstants(Function *F,
...
+      float value;
+      if (getFloatScalar(mulNode->getRHS(), &value)) {
+        otherMulOperand = maxNode->getLHS();
+      } else if (getFloatScalar(mulNode->getLHS(), &value)) {
+        otherMulOperand = maxNode->getRHS();
+      } else {
+        continue;
+      }
+      if ((value <= 1.0f) && (otherMulOperand == otherMaxOperand)) {
+        // The sub-tree is a Leaky-Relu, express it as a PRelu
```
[the remark this answers is not in the document]
Comment (tlepley-cadence) [reply lib/Optimizer/GraphOptimizer.cpp]: Done.

File: lib/Optimizer/GraphOptimizer.cpp [hunk 7]
```diff
@@ -2453,6 +2453,86 @@ void glow::convertPlaceholdersToConstants(Function *F,
   }
 }
 
+/// Return true if the \p node sub-tree corresponds to a scalar
```
[the remark this answers is not in the document]
Comment (tlepley-cadence) [reply lib/Optimizer/GraphOptimizer.cpp]: Done.

File: lib/Optimizer/GraphOptimizer.cpp [hunk 8]
```diff
@@ -2453,6 +2453,86 @@ void glow::convertPlaceholdersToConstants(Function *F,
...
+/// (Constant or Splat) and return the float value in \p retFloat.
+static bool getFloatScalar(Node *node, float *retFloat) {
+  // Iterate accross potential Tile Nodes (implied by broadcasting if any).
```
[the remark this answers is not in the document]
Comment (tlepley-cadence) [reply lib/Optimizer/GraphOptimizer.cpp]: Done.

**Comment by tlepley-cadence** [comment]
@jfix71 All done.
